components/use-vault.js

```diff
@@ -0,0 +1,313 @@
+      if (data?.setVaultEntry) {
+        unsetLocalStorage(storageKey)
+        window.localStorage.removeItem(storageKey.localKey)
+        migratedCount++
+        console.log('stored encrypted value in vault:', storageKey, value, encrypted)
+      } else {
+        console.error('failed to store encrypted value in vault:', storageKey)
+      }
+    }
+
+    return migratedCount
+  }, [me?.id])
+
+  return migrate
+}
+
+// used to get and set values in the vault
+export default function useVault (unscopedStorageKey, defaultValue) {
+  const { me } = useMe()
+
+  // scope all storage to user
+  const storageKey = `${unscopedStorageKey}:${me?.id}`
+
+  const localOnly = storageKey.includes(':local-only:')
+
+  const [setVaultValue] = useMutation(SET_ENTRY)
+  const [value, innerSetValue] = useState(undefined)
+  const [clearVaultValue] = useMutation(UNSET_ENTRY)
+  const { data: vaultData, refetch: refetchVaultValue } = useQuery(GET_ENTRY, {
+    variables: { key: storageKey },
+    // fetchPolicy only applies to first execution on mount so we also need to
+    // set nextFetchPolicy to make sure we don't serve stale values from cache
+    nextFetchPolicy: 'no-cache',
+    fetchPolicy: 'no-cache'
+  })
+
+  useEffect(() => {
+    (async () => {
+      if (!me) return
+
+      if (localOnly) {
+        innerSetValue(getLocalStorage(storageKey) || defaultValue)
+        return
+      }
+
+      const localVaultKey = getLocalKey(me.id)
+
+      if (!me.privates.vaultKeyHash || localVaultKey?.hash !== me.privates.vaultKeyHash) {
+        // no or different vault setup on server
+        // use unencrypted local storage
+        unsetLocalKey(me.id)
+        innerSetValue(getLocalStorage(storageKey) || defaultValue)
+        return
+      }
+
+      // if vault key hash is set on the server, vault entry exists and vault key is set on the device
+      // decrypt and use the value from the server
+      const encrypted = vaultData?.getVaultEntry?.value
+      if (encrypted) {
+        try {
+          const decrypted = await decryptJSON(localVaultKey.key, encrypted)
+          // console.log('decrypted value from vault:', storageKey, encrypted, decrypted)
+          innerSetValue(decrypted)
+          // remove local storage value if it exists
+          unsetLocalStorage(storageKey)
+          return
+        } catch (e) {
+          console.error('cannot read vault data:', storageKey, e)
+        }
+      }
+
+      // fallback to local storage
+      innerSetValue(getLocalStorage(storageKey) || defaultValue)
+    })()
+  }, [vaultData, me?.privates?.vaultKeyHash])
+
+  const setValue = useCallback(async (newValue) => {
+    const vaultKey = getLocalKey(me.id)
+
+    const useVault = vaultKey && vaultKey.key && vaultKey.hash === me.privates.vaultKeyHash
+
+    if (useVault && !localOnly) {
+      const encryptedValue = await encryptJSON(vaultKey.key, newValue)
+      await setVaultValue({ variables: { key: storageKey, value: encryptedValue } })
```

Comment:
it shouldn't use the scoped key here, the unique constraint is userId+key in the database

Reply:
But is it a problem?

Reply:
It could be because if you have to change the way the key is scoped in future, you will have to migrate every key in the vault. Since we don't need this additional information in the key, it is better to not add it, imo.

Reply:
You reminded me that I've actually been thinking about hashing the keys so we don't even know what the users are storing. The names leak metadata.